[DOC] Add Object Detection Explanation (#1807)
open-edge-platform/training_extensions: docs/source/guide/explanation/algorithms/object_detection/object_detection.rst

Object Detection
================

Object detection is a computer vision task where it's needed to locate objects, finding their bounding boxes coordinates together with defining class. 

> The input is an image, and the output is a pair of coordinates for bouding box corners with the number of class for each detected object.

Reviewer: ```suggestion
The input is an image, and the output is a pair of coordinates for bouding box corners with classes for each detected object.
```

      
      .. code-block::

        $ otx train {TEMPLATE} ... \
            params \
            --learning_parameters.enable_early_stopping=True

    - `Anchor clustering for SSD <https://arxiv.org/abs/2211.17170>`_: This model highly relies on predefined anchor boxes hyperparameter that impacts the size of objects, which can be detected. So before training, we collect object statistics within dataset, cluster them and modify anchor boxes sizes to fit the most for objects the model is going to detect. 
    
    - ``Backbone pretraining``: we pretrained MobileNetV2 backbone on large `ImageNet21k <https://github.com/Alibaba-MIIL/ImageNet21K>`_ dataset to improve feature extractor and learn better and faster.

**************
Dataset Format
**************

At the current point we support `COCO <https://cocodataset.org/#format-data>`_, 
`Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_ and
`YOLO <https://openvinotoolkit.github.io/datumaro/docs/formats/yolo/>`_ dataset format.
The dataset should be formatted according to the structure below:

> .. code::

Reviewer: Is this structure same for all dataset format(including VOC, YOLO)? If not, this part should be added a explanation that this is example for COCO dataset format.

Reviewer 2: Nice comment @jaegukhyun , we should differentiate the dataset format

Author: Thank you for this comments, I mentioned COCO format

  ├── annotations/
      ├── instances_train.json
      ├── instances_val.json
      └── instances_test.json
  ├──images/
      (Split is optional)
      ├── train
      ├── val
      └── test

If you have your dataset in those formats, then you can simply run using one line of code:

.. code::

>     $ otx train ATSS --train-data-root <path_to_data_root> \

Reviewer: ```suggestion
    $ otx train ATSS --train-data-roots <path_to_data_root> \
```

>                      --val-data-root <path_to_data_root>

Reviewer: ```suggestion
                     --val-data-roots <path_to_data_root>
```

+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+---------------------+-----------------+
| `Custom_Object_Detection_Gen3_ATSS <https://github.com/openvinotoolkit/training_extensions/blob/develop/otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml>`_  | ATSS  | 20.6                | 9.1             |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+---------------------+-----------------+

`ATSS <https://arxiv.org/abs/1912.02424>`_ is a good medium-range model that works well and fast in most cases. 
`SSD <https://arxiv.org/abs/1512.02325>`_ and `YOLOX <https://arxiv.org/abs/2107.08430>`_ are light models, that a perfect for the fastest inference on low-power hardware.

> YOLOX achieved the same accuracy as SSD, and even outperforms its inference on CPU 1.5 times, but requires 3 times more time for training, which is even more than for ATSS. So if you have resources for a long training, you can pick the YOLOX model.

Reviewer: YOLOX has fast inference speed but requires more training time due to Mosaic augmentation. I think this information may help for users understanding

Author: Thank you, added this and link with Mosaic augmentation

Besides this, we support public backbones from `torchvision <https://pytorch.org/vision/stable/index.html>`_, `pytorchcv <https://github.com/osmr/imgclsmob>`_, `mmcls <https://github.com/open-mmlab/mmclassification>`_ and `OpenVino Model Zoo <https://github.com/openvinotoolkit/open_model_zoo>`_.
Please, refer to the :doc:`tutorial <../../../tutorials/advanced/backbones>` how to customize models and run public backbones.

To see which public backbones are available for the task, the following command can be executed:

.. code-block::

        $ otx find --backbone {torchvision, pytorchcv, mmcls, omz.mmcls}

In the table below the mAP on some academic datasets using our :ref:`supervised pipeline <od_supervised_pipeline>` is presented.
The results were obtained on our templates without any changes.
For hyperparameters, please, refer to the related template.
We trained each model with a single Nvidia GeForce RTX3090.

+-----------+------------+-----------+
| Model name| COCO       | MinneApple|
+===========+============+===========+

> | YOLOX     | N/A        | N/A       |

Reviewer: Will N/A part be added in this PR?

Author: Good point. For COCO I don't have results on hand, but I found results for MinneApple (although it's not a well-known dataset)

Author: Added Pascal VOC dataset to be trained as well

Reviewer: I think we can add pretrained weights' performance for COCO or we can delete the COCO performance in this table.
I think deleting could be an option since OTX is transfer learning tool(COCO -> COCO looks weird). In addition VOC results should be contained in this table. Will you update this?

Author: Sadly, I don't have these numbers for Pascal VOC, there is no way to keep it as N/A? I have an accuracy of pretrained weights for ATSS and SSD, but not for YOLOX.
It's better to keep a half-filled table or delete it all?
